src/main/java/graphql/nadel/engine/OverallQueryTransformer.java

```diff
@@ -444,7 +450,44 @@ public TraversalControl enter(TraverserContext<Node> context) {
         );
 
         //noinspection unchecked
-        return (T) newNode;
+        return (T) addUnderscoreTypeNameToEmptySelectionSets(nadelContext, overallTypeInformation, rootVars, treeTransformer, newNode);
```

Comment:
Trying to traverse the tree again after we traversed it the first time. @andimarek as you correctly said, during the first traversal when we are inside of a parent node we don't know whether the children are going to be deleted or not so we can't make a correct decision on whether the typename alias should be added. During the second traversal we do have this information so we can add the dummy field conditionally. 

Not sure whether this approach is acceptable as we are going to traverse the whole tree again. I think we can optimize the second traversal a bit. During the first traversal when leaving a node we can check whether all children were deleted. We can collect all nodes like that into a set and during the second traversal instead of checking every time whether the alias needs to be added we can just look it up from the set?

I also considered another approach. I though about adding the typename alias when leaving the node during the first traversal by overriding the `leave` method [here](https://github.com/graphql-java/nadel/blob/master/src/main/java/graphql/nadel/engine/OverallQueryTransformer.java#L438) but looks like we can't modify the node on leave. Not sure whether we want to change that?

Reply:
> I think we can optimize the second traversal a bit.

that probably not going to do anything though :(